src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,273 @@
+
+        internal const int MaxByteBufferSize = 4096;
+        internal const int MaxCharBufferSize = 3 * MaxByteBufferSize;
+
+        private readonly Stream _stream;
+        private readonly Decoder _decoder;
+
+#if NETSTANDARD2_0
+        private readonly Encoder _encoder;
+#endif
+
+        private ArraySegment<byte> _byteBuffer;
+        private ArraySegment<char> _charBuffer;
+        private ArraySegment<byte> _overflowBuffer;
+        private bool _disposed;
+
+        public TranscodingReadStream(Stream input, Encoding sourceEncoding)
+        {
+            _stream = input;
+
+            // The "count" in the buffer is the size of any content from a previous read.
+            // Initialize them to 0 since nothing has been read so far.
+            _byteBuffer = new ArraySegment<byte>(
+                ArrayPool<byte>.Shared.Rent(MaxByteBufferSize),
+                0,
+                count: 0);
+
+            // Attempt to allocate a char buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the byte -> char conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            // The conversion process is tolerant of char buffer that is not large enough to convert all the bytes at once.
+            int maxCharBufferSize = Math.Min(MaxCharBufferSize, sourceEncoding.GetMaxCharCount(MaxByteBufferSize));
+            _charBuffer = new ArraySegment<char>(
+                ArrayPool<char>.Shared.Rent(maxCharBufferSize),
+                0,
+                count: 0);
+
+            _overflowBuffer = new ArraySegment<byte>(
+                ArrayPool<byte>.Shared.Rent(OverflowBufferSize),
+                0,
+                count: 0);
+
+            _decoder = sourceEncoding.GetDecoder();
+
+#if NETSTANDARD2_0
+            _encoder = Encoding.UTF8.GetEncoder();
+#endif
+        }
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => false;
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        internal int ByteBufferCount => _byteBuffer.Count;
+        internal int CharBufferCount => _charBuffer.Count;
+        internal int OverflowCount => _overflowBuffer.Count;
+
+        public override void Flush()
+            => throw new NotSupportedException();
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowArgumentOutOfRangeException(buffer, offset, count);
```

Comment:
We generally do such argument validation outside of async methods, as otherwise the exception will end up being stored into the returned task rather than propagated out to the synchronous caller, e.g.
```C#
public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
{
    ValidateArguments(buffer, offset, count);
    return ReadAsyncCore(buffer, offset, count, cancellationToken);
}

private async Task<int> ReadAsyncCore(...) { ... }
```

Reply:
I don't think we can move this validation since the caller for this method is always going to be the `JsonSerializer.SerializeAsync` method in System.Text.Json library.

Reply:
I'm not sure why we couldn't do what I outlined above; it doesn't matter who the caller is, it's an implementation detail of the ReadAsync method.  That said, if the only caller is our own async method, the change isn't worthwhile.

Reply:
nvm, you are right.